design/baremetal-operator/configmap_for_host_inspection_data.md

```diff
@@ -0,0 +1,193 @@
+else will be given no write permissions. If the BareMetalHost was inspected but
+HostData object doesn't exist, operator will first try to copy it from the
+BareMetalHost status (if available) and vice versa. If not available in neither
+of them, new inspection will be enforced.
+
+#### Keeping inspection data in sync
+
+As mentioned above, we will have two copies of inspection data available. One in the Spec
+of HostData and the other in the Status of BareMetalHost. One of them will get its inspection
+data copied from the other. Thus, inspection data can be copied from HostData (if available)
+to BareMetalHost, or the vice versa. If there is delta in the inspection data while
+copying, HostData should take precedence and get copied to the Status of BareMetalHost.
+
+**Note:** [cross-namespace owner references are not allowed](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications). Thus, HostData
+and BareMetalHost must be in the same namespace.
+
+- BareMetalHost is created
+- BareMetalHost is in inspecting state
+- Operator creates a HostData with the same name and namespace as BareMetalHost, fills out
+    the HostData with the host inspection data and sets ownerReferences
+- Provisioning is requested
+- Pivoted to a target cluster, operator restores host inspection data from HostData and
+    the rest from status annotation (e.g. `status.Provisioning`, `status.poweredOn` and etc)
+- Deprovisioning is requested
+- BareMetalHost deletion requested, operator deletes the HostData and BareMetalHost
+
+The spec of HostData:
+
+```go
+// Adding HardwareDetails struct here for visibility. During implementation,
+// we don't have to duplicate it, but use from baremetalhost_types.go.
+type HardwareDetails struct {
+
+    SystemVendor HardwareSystemVendor `json:"systemVendor,omitempty"`
+    Firmware     Firmware             `json:"firmware,omitempty"`
+    RAMMebibytes int                  `json:"ramMebibytes,omitempty"`
+    NIC          []NIC                `json:"nics,omitempty"`
+    Storage      []Storage            `json:"storage,omitempty"`
+    CPU          CPU                  `json:"cpu,omitempty"`
+    Hostname     string               `json:"hostname,omitempty"`
+
+}
+
+type HostDataSpec struct {
+    // The hardware discovered on the host.
+    HardwareDetails *HardwareDetails `json:"hardware,omitempty"`
+}
+```
+
+### Risks and Mitigations
+
+Existing deployments will require installation of the HostData CRD and necessary
+RBAC permissions when bumping the operator version. This needs to be clearly
+documented.
+
+### Work Items
+
+- Add HostData API
+- Modify the operator code
+- Adjust CAPM3 machine controller to exlude writing `status.hardware` into the statusAnnotation
+- Add unit tests
+- Update documentation
+
+### Dependencies
+
+None
+
+### Test Plan
+
+Unit tests should be added. Integration test doesn't require any change.
```

Comment:
AFAIK we don't yet have e2e in BMO. If there is one in the future I can add it, but perhaps for now this can be continued to be tested in daily master jobs (Jenkins) that use metal3-dev-env code base?

Reply:
That's right, in any case I think there could be some value to define, even in a pure natural language, which e2e test could make sense (or it's reasonable) to have for validating/accepting the new feature - so that maybe it could be easier in future to implement them having previously defined in the design doc. I wasn't pointing to an exhaustive or super-detailed list, just a few of them